### PR TITLE
Fix bug with creating new object while editing

### LINKED ISSUE #4545: Admin create new entity after ModelManagerException instead of update existing entity
#### Sonata packages

```bash
sonata-project/admin-bundle              3.4.0  The missing Symfony Admin ...
sonata-project/block-bundle              3.1.1  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/core-bundle               3.0.3  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2    Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  Doctrine2 behavioral exten...
sonata-project/doctrine-orm-admin-bundle 3.0.5  Symfony Sonata / Integrate...
sonata-project/easy-extends-bundle       2.1.10 Symfony SonataEasyExtendsB...
sonata-project/exporter                  1.7.0  Lightweight Exporter library
sonata-project/google-authenticator      1.0.2  Library to integrate Googl...
sonata-project/user-bundle               3.0.1  Symfony SonataUserBundle
```

## Subject
If in moment of entity update happens sql error (in our case doctrine throws Doctrine\DBAL\Exception\UniqueConstraintViolationException), admin edit page shows with error and form, but form have url of creating new objet instead of edit.
Method Sonata\AdminBundle\Controller\CRUDController->editAction

## Steps to reproduce
Make Admin class inherited from AbstractAdmin and uses CRUDController
Crate several new object
Open edit page
Fill OneToOne field with value used by other object
Click Update
See error "An error has occurred during update of item.."
Fill OneToOne field with null value or unused value
Click Create

## Expected results
See edit form after error on update

## Actual results
In fact then DBAL throws UniqueConstraintViolationException UnitOfWork catch exception and close EntityManager, which execute clearing of identityMap.
And when admin class start building template, it use admin.id(object) (in base_edit_form.html.twig)
who asks to ModelManager, which look in identityMap. But in identy map we don't have this object, because map is cleared after object is fetched from database. And this object for template look like new.

### PR DESCRIPTION
**Pull request by UnDeleteRU**
Closes #4545

## Changelog

```markdown
### Changed
- Passing object id in edit form from CRUD controller instead of getting it in twig
```

## Subject
Passing object id from controller prevent situation when in twig we can't get object id because model manager reopened. It geting id before starts operations with database and will always open edit form even we have some database exception.

### CODE REVIEW

**Comment by greg0ire** [comment]
> because model manager reopened

Not sure what you mean by that. Is the issue a lack of clear exception?

**Comment by jlamur** [comment]
@greg0ire `AbstractAdmin::id(...)` throws an exception when the EntityManager is closed. The idea of this PR is to get the id from the EntityManager before trying to execute an update (which could close the EntityManager).

**Comment by greg0ire** [comment]
Ok I think I understand. Suggested commit message:

> Fetch the id before attempting any update
>
> A failing update is hard to recover from and the id cannot be retrieved when
the entity manager is closed.
> Fixes #4545

File: Resources/views/CRUD/base_edit.html.twig [hunk 1]
```diff
@@ -12,7 +12,7 @@ file that was distributed with this source code.
 {% extends base_template %}
 
 {% block title %}
-    {% if admin.id(object) is not null %}
+    {% if objectId is not null %}
```
Comment (greg0ire) [inline Resources/views/CRUD/base_edit.html.twig]: This is a BC-break for people that extend the `CrudController`. A solution could be this:

`{% if objectId|default(admin.id(object)) is not null %}`

**Comment by UnDeleteRU** [comment]
I fix commit message and BC compability

File: Resources/views/CRUD/base_edit.html.twig [hunk 2]
```diff
@@ -12,7 +12,7 @@ file that was distributed with this source code.
 {% extends base_template %}
 
 {% block title %}
-    {% if admin.id(object) is not null %}
+    {% if objectId|default(admin.id(object)) is not null %}
```
Comment (greg0ire) [inline Resources/views/CRUD/base_edit.html.twig]: I forgot: please also add {# NEXT_MAJOR: remove default filter #}

Comment (UnDeleteRU) [inline Resources/views/CRUD/base_edit.html.twig]: I add it only at base_edit.html.twig, it's enough? or i need add it in base_edit_form,html.twig also?

Comment (greg0ire) [inline Resources/views/CRUD/base_edit.html.twig]: Add it everywhere please. Act as if a five year old was doing the next major cleanup.

Comment (UnDeleteRU) [inline Resources/views/CRUD/base_edit.html.twig]: Okay, add comment to all places.

**Comment by greg0ire** [comment]
Also, sorry for the broken CI, you will have to wait for https://github.com/sonata-project/dev-kit/pull/257 to be deployed on this repo, then simply close and reopen your PR to trigger the build. should happen this night.

**Comment by greg0ire** [comment]
Restarted the build

**Comment by UnDeleteRU** [comment]
@greg0ire i fix test and make all requested changes. Anything else i need to do?

**Comment by greg0ire** [comment]
No, it's all good. @sonata-project/contributors please review

File: Controller/CRUDController.php [hunk 3]
```diff
@@ -270,11 +271,12 @@ public function editAction($id = null)
 
                 try {
                     $existingObject = $this->admin->update($submittedObject);
+                    $objectId = $this->admin->getNormalizedIdentifier($existingObject);
```
Comment (jlamur) [inline Controller/CRUDController.php]: Why do set `$objectId` again here?

File: Resources/views/CRUD/base_edit_form.html.twig [hunk 4]
```diff
@@ -2,17 +2,19 @@
...
         <div>
             {{ "form_not_available"|trans({}, "SonataAdminBundle") }}
         </div>
     {% else %}
+        {# NEXT_MAJOR: remove default filter #}
         <form
               {% if sonata_admin.adminPool.getOption('form_type') == 'horizontal' %}class="form-horizontal"{% endif %}
               role="form"
-              action="{% block sonata_form_action_url %}{{ admin.generateUrl(url, {'id': admin.id(object), 'uniqid': admin.uniqid, 'subclass': app.request.get('subclass')}) }}{% endblock %}"
+              action="{% block sonata_form_action_url %}{{ admin.generateUrl(url, {'id': objectId|default(admin.id(object)), 'uniqid': admin.uniqid, 'subclass': app.request.get('subclass')}) }}{% endblock %}"
```
Comment (jlamur) [inline Resources/views/CRUD/base_edit_form.html.twig]: Please put the `NEXT_MAJOR` comment right above of this line.

Comment (UnDeleteRU) [inline Resources/views/CRUD/base_edit_form.html.twig]: fix this

File: Controller/CRUDController.php (hunk 3 above)
Comment (UnDeleteRU) [inline Controller/CRUDController.php]: When i see this code first time, i think that object id can change while update.
I fix this.

**Comment by jlamur** [comment]
@UnDeleteRU please fix StyleCI

**Comment by OskarStark** [comment]
@UnDeleteRU after the fix, this one could be merged!

**Comment by UnDeleteRU** [comment]
@jlamur @OskarStark i fix StyleCI

**Comment by OskarStark** [comment]
Thank you @UnDeleteRU 

Please rebase, we fixed the travis configuration for the PHP 5.3 builds, thank you

**Comment by greg0ire** [comment]
No need for a rebase, let's just trigger the build again

**Pull request merged**